pkg/controller/oam/applicationconfiguration/render.go

```diff
@@ -106,13 +119,88 @@ func (r *components) Render(ctx context.Context, ac *v1alpha2.ApplicationConfigu
 			t.SetNamespace(ac.GetNamespace())
 
 			traits[i] = *t
-		}
 
-		workloads[i] = Workload{ComponentName: acc.ComponentName, Workload: w, Traits: traits}
+			traitDefinitionName := getCRDName(t)
+			var traitDef v1alpha2.TraitDefinition
+			err = r.client.Get(ctx, types.NamespacedName{
+				Name: traitDefinitionName,
+			}, &traitDef)
+			if err != nil {
+				return nil, errors.Wrapf(err, errFmtGetTraitDefinition, traitDefinitionName)
+			}
+			traitDefs = append(traitDefs, traitDef)
+		}
+		if err := SetWorkloadInstanceName(traitDefs, w, c); err != nil {
+			return nil, err
+		}
+		workloads[i] = Workload{ComponentName: acc.ComponentName, RevisionName: acc.RevisionName, Workload: w, Traits: traits}
 	}
 	return workloads, nil
 }
 
+// SetWorkloadInstanceName will set metadata.name for workload CR according to createRevision flag in traitDefinition
+func SetWorkloadInstanceName(traitDefs []v1alpha2.TraitDefinition, w *unstructured.Unstructured, c *v1alpha2.Component) error {
+	//Don't override the specified name
+	if w.GetName() != "" {
```

Comment:
IIRC, a user can actually override the "CreateRevision" field in the trait? I am not sure if that's ok

Reply:
No, a user can't, CreateRevision field is defined in traitDefinition which won't  be changed by user